src/scanner.ts

```diff
@@ -646,45 +646,53 @@ export class Scanner {
         let start = this.pos;
         while (true) {
             if (this.pos >= this.len) {
-                result += this.text.substring(start, this.pos);
+                result += this.text.slice(start, this.pos);
                 this.setTokenAsUnterminated();
                 this.getDiagnostics().report(this.pos, diagnostic || Diagnostics.Unterminated_string_literal);
                 break;
             }
 
             let ch = this.text.charCodeAt(this.pos);
+            let chFromEntity = false;
             if (decodeEscapeSequences && ch === CharacterCodes.Ampersand) {
-                result += this.text.substring(start, this.pos);
+                result += this.text.slice(start, this.pos);
                 this.pos++;
                 ch = this.scanCharacterEntity();
+
+                // Decoding escape sequences breaks assumptions about character start position.
+                // Set pos to the end of the escape sequence but slice start one character later,
+                // and record that we are in such a state.
                 start = this.pos;
+                this.pos--;
+                chFromEntity = true;
             }
 
             if (ch === quote) {
                 // If this is a terminal that consists solely of a single backtick character (e.g. ```),
                 // we capture the backtick.
-                if (quote === CharacterCodes.Backtick && this.pos === start && this.pos + 1 < this.len) {
+                if (quote === CharacterCodes.Backtick && result === "" && this.pos + 1 < this.len) {
                     ch = this.text.charCodeAt(this.pos + 1);
                     if (ch === CharacterCodes.Backtick) {
-                        result = "`";
-                        this.pos += 2;
-                        break;
+                        this.pos++;
+                        if (chFromEntity) {
+                            result = "`";
+                        }

```

Comment:
Is there a reason this no longer has a `break`? This should consume `` ``` `` and then `break` (which returns the result of `` ` ``), however it looks like it will now concatenate additional characters because it falls through to line 683.

Reply:
The characters that get concatenated are in fact a single backtick... this new code is leveraging the redundancy. AFAICT, there is no change in that case:

1. initial state: <code>this.text = "```…"</code>, `this.pos = 0`
2. L645: `result = ""`
3. L646: `start = 0`
4. L655: <code>ch = "`"</code> (the first backtick)
5. L674: <code>ch = "`"</code> (the second backtick)
6. L676: `this.pos = 1`
7. L683: <code>result += "`"</code> (the first backtick, not that it matters)
8. L684: `this.pos = 2`
9. L685: `break`
10. L704: <code>return "`"</code>

Reply:
I need to think about how the decoding rules should be handled for ` ``` ` when HTML entities are involved. For example,
should all of the following cases be treated the same as ` ``` `:

- ```` &#x60;`` ````
- ```` `&#x60;` ````
- ```` ``&#x60; ````
- ```` &#x60;&#x60;` ````
- ```` &#x60;`&#x60; ````
- etc.

Since we treat `&#x60;` as a non-terminal quote character, then we need to further handle the entity for the end of the quote, otherwise we will prematurely end the nonterminal for ```` &#x60;`&#x60; ```` at the unencoded backtick in the middle.

Reply:
I'm of the opinion that <code>&#x60;``</code> should _not_ be treated the same as raw <code>```</code>, but that was already the case so I preserved it because it is not part of the issue at hand. It seems like something to fix later.

Reply:
Grammarkdown itself doesn't care about entity escapes, support for them was added to get accurate line/column positions for diagnostics and text editor features when a grammarkdown grammar is embedded in an ECMArkup document. Prior to this we were just unescaping the entities ahead of time, which resulted in incorrect positions.

Reply:
Its a corner case, since in general we'd be unlikely to encounter an escaped backtick, so I may accept this change until such time as I can add a preprocessor for entity decoding.

Reply:
Ah, I think I understand. Let me tweak things a little bit more.

Reply:
Done. I updated the <code>\`\`\`</code> peek logic such that <code>``&amp;#x60;</code> and <code>\`&amp;#x60;`</code> are treated identically.